README.md
# pg-copy-binary

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar pg-copy-binary-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-04-21
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-04-21
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/pg-copy-binary/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/pg-copy-binary/compare/0.1.0...0.1.1


test/pg_copy_binary/core_test.clj
(ns pg-copy-binary.core-test
  (:require [clojure.test :refer :all]
            [pg-copy-binary.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/pg_copy_binary/encode.clj
(ns pg-copy-binary.encode
  (:require
   [pg.oid :as oid])
)


(defn arr16 ^bytes [value]
  (byte-array
   [(-> value (bit-and 0xff00) (bit-shift-right 8))
    (-> value (bit-and 0x00ff))]))


(defn arr32 ^bytes [value]
  (byte-array
   [(-> value (bit-and 0xff000000) (bit-shift-right 24))
    (-> value (bit-and 0x00ff0000) (bit-shift-right 16))
    (-> value (bit-and 0x0000ff00) (bit-shift-right  8))
    (-> value (bit-and 0x000000ff))]))


(defn arr64 ^bytes [value]
  (let [buf
        (-> value
            (BigInteger/valueOf)
            (.toByteArray))

        pad
        (- 8 (alength buf))]

    (if (pos? pad)
      (byte-array (-> []
                      (into (repeat pad 0))
                      (into buf)))
      buf)))


(defmulti -encode
  (fn [value oid _]
    [(type value) oid]))


(defmethod -encode [String oid/TEXT]
  [^String value _ _]
  (.getBytes value "UTF-8"))


(defmethod -encode [Character oid/TEXT]
  [^Character value oid opt]
  (-encode (str val) oid opt))


;;
;; Long
;;

(defmethod -encode [Long oid/INT8]
  [^Long value _ _]
  (arr64 value))


(defmethod -encode [Long oid/INT4]
  [^Long value oid opt]
  (-encode (int value) oid opt))


(defmethod -encode [Long oid/INT2]
  [^Long value oid opt]
  (-encode (short value) oid opt))


;;
;; Integer
;;

(defmethod -encode [Integer oid/INT8]
  [^Integer value oid opt]
  (-encode (long value) oid opt))


(defmethod -encode [Integer oid/INT4]
  [^Integer value oid opt]
  (arr32 value))


(defmethod -encode [Integer oid/INT2]
  [^Integer value oid opt]
  (-encode (short value) oid opt))


;;
;; Short
;;

(defmethod -encode [Short oid/INT8]
  [^Short value oid opt]
  (-encode (long value) oid opt))


(defmethod -encode [Short oid/INT4]
  [^Short value oid opt]
  (-encode (int value) oid opt))


(defmethod -encode [Short oid/INT2]
  [^Short value oid opt]
  (arr16 value))


;;
;; Bool
;;

(defmethod -encode [Boolean oid/BOOL]
  [^Boolean value _ _]
  (let [b
        (case value
          true 1
          false 0)]
    (byte-array [b])))


;;
;; Float
;;

(defmethod -encode [Float oid/FLOAT4]
  [^Float value oid opt]
  (-> (Float/floatToIntBits value)
      (arr32)))


(defmethod -encode [Float oid/FLOAT8]
  [^Float value oid opt]
  (-encode (double value) oid opt))


;;
;; Double
;;

(defmethod -encode [Double oid/FLOAT4]
  [^Double value oid opt]
  (-encode (float value) oid opt))


(defmethod -encode [Double oid/FLOAT8]
  [^Double value oid opt]
  (-> (Double/longBitsToDouble value)
      (arr64)))


(def defaults
  {Double oid/FLOAT8
   Float  oid/FLOAT4
   String oid/TEXT
   Long   oid/INT8

   }
  )


src/pg_copy_binary/joda_time.clj
(ns pg-copy-binary.joda-time
  (:require
   [pg-copy-binary.core :as core])
  (:import
   org.joda.time.DateTimeZone
   org.joda.time.Instant
   org.joda.time.Days
   org.joda.time.LocalTime
   org.joda.time.Time ;;
   org.joda.time.Date ;;
   org.joda.time.LocalDate
   org.joda.time.DateTime
   org.joda.time.LocalDateTime))


;;
;; Const
;;

(def ^LocalDate LD_EPOCH
  (new LocalDate 1970 1 1))


(extend-protocol core/IPGWrite

  LocalTime

  (-bytes [this]
    (core/arr64 (.getMillisOfDay this)))

  LocalDate

  (-bytes [this]
    (let [days
          (-> (Days/daysBetween LD_EPOCH this)
              (.getDays)
              (- (.toDays core/PG_EPOCH_DIFF)))]
      (core/arr32 days)))

  LocalDateTime

  (-bytes [this]
    (core/-bytes (.toDateTime this DateTimeZone/UTC)))

  DateTime

  (-bytes [this]

    (let [millis
          (-> this .getMillis )

          sec-millis
          (.getMillisOfSecond this)]

      (core/arr64
       (+ (- millis (.toMillis core/PG_EPOCH_DIFF))
          (* sec-millis 1000))))))


src/pg_copy_binary/core.clj
(ns pg-copy-binary.core
  (:require
   [clojure.java.io :as io])

  (:import
   java.util.Date
   java.util.UUID

   java.time.Instant
   java.time.Duration
   java.time.LocalDate
   java.time.LocalTime
   java.time.ZoneOffset

   java.io.InputStream
   java.io.ByteArrayOutputStream
   java.io.OutputStream
   java.io.Writer))


(def ^Duration PG_EPOCH_DIFF
  (Duration/between Instant/EPOCH
                    (-> (LocalDate/of 2000 1 1)
                        (.atStartOfDay)
                        (.toInstant ZoneOffset/UTC))))


(defn arr16 ^bytes [value]
  (byte-array
   [(-> value (bit-and 0xff00) (bit-shift-right 8))
    (-> value (bit-and 0x00ff))]))


(defn arr32 ^bytes [value]
  (byte-array
   [(-> value (bit-and 0xff000000) (bit-shift-right 24))
    (-> value (bit-and 0x00ff0000) (bit-shift-right 16))
    (-> value (bit-and 0x0000ff00) (bit-shift-right  8))
    (-> value (bit-and 0x000000ff))]))


(defn arr64 ^bytes [value]
  (let [buf
        (-> value
            (BigInteger/valueOf)
            (.toByteArray))

        pad
        (- 8 (alength buf))]

    (if (pos? pad)
      (byte-array (-> []
                      (into (repeat pad 0))
                      (into buf)))
      buf)))


;;
;; Const
;;

(def ^{:tag 'bytes} HEADER
  (byte-array 11 (map int [\P \G \C \O \P \Y \newline 0xFF \return \newline 0])))


(def ^{:tag 'bytes} zero32
  (arr32 0))


(def ^{:tag 'bytes} -one32
  (arr32 -1))


(def ^{:tag 'bytes} -one16
  (arr16 -1))


(defprotocol IPGWrite
  (-bytes ^bytes [this]))


(extend-protocol IPGWrite

  String

  (-bytes [this]
    (.getBytes this "UTF-8"))

  Character

  (-bytes [this]
    (-bytes (str this)))

  clojure.lang.Symbol

  (-bytes [this]
    (-bytes (str this)))

  clojure.lang.Keyword

  (-bytes [this]
    (-bytes (-> this str (subs 1))))

  UUID

  (-bytes [this]

    (let [most-bits
          (.getMostSignificantBits this)

          least-bits
          (.getLeastSignificantBits this)]

      (byte-array
       (-> []
           (into (arr64 most-bits))
           (into (arr64 least-bits))))))

  LocalDate

  (-bytes [this]
    (arr32
     (- (.toEpochDay this)
        (.toDays PG_EPOCH_DIFF))))

  LocalTime

  (-bytes [this]
    (arr64 (quot (.toNanoOfDay this) 1000)))

  Instant

  (-bytes [this]

    (let [seconds
          (- (.getEpochSecond this)
             (.toSeconds PG_EPOCH_DIFF))

          nanos
          (.getNano this)]

      (arr64
       (+ (* seconds 1000 1000)
          (.getNano this)))))

  Date

  (-bytes [this]
    (-bytes (.toInstant this)))

  Boolean

  (-bytes [this]
    (let [b
          (case this
            true 1
            false 0)]
      (byte-array [b])))

  Short

  (-bytes [this]
    (arr16 this))

  Integer

  (-bytes [this]
    (arr32 this))

  Long

  (-bytes [this]
    (-> (BigInteger/valueOf this)
        (arr64)))

  Float

  (-bytes [this]
    (-> (Float/floatToIntBits this)
        (arr32)))

  Double

  (-bytes [this]
    (-> (Double/longBitsToDouble this)
        (arr64))))


(defn table->out [table ^OutputStream out]
  (.write out HEADER)
  (.write out zero32)
  (.write out zero32)
  (doseq [row table]
    (.write out (arr16 (count row)))
    (doseq [item row]
      (if (nil? item)
        (.write out -one32)
        (let [buf
              (-bytes item)]
          (.write out (arr32 (alength buf)))
          (.write out buf)))))
  (.write out -one16)
  (.close out))


(defn table->bytes ^bytes [table]
  (with-open [out (new ByteArrayOutputStream)]
    (table->out table out)
    (.toByteArray out)))


(defn table->file [table ^String path]
  (with-open [out (-> path
                      io/file
                      io/output-stream)]
    (table->out table out)))


(defn table->input-stream ^InputStream [table]
  (-> table
      table->bytes
      io/input-stream))


(defn maps->table [maps row-keys]
  (map (apply juxt row-keys) maps))
